### ISSUE TITLE
Use local storage instead of hostPath on K8s 1.7+

### ISSUE THREAD

**Issue by travisn** [rook/rook#796]
K8s 1.7 comes with the new local storage feature, which is a much-needed improvement over the hostPath. This would replace the need for the `dataDirHostPath` in the cluster yaml. Need to look at if we can always use local storage automatically, though still allow setting the path in the cluster yaml.

**Comment by jbw976** [comment]
There is potentially some overlap with #726 where perhaps all pod configuration could be stored in configmaps or other kubernetes facilities, instead of relying on ceph config files to be persisted on the host of the pod using the new "local storage" feature.

Note that wouldn't solve the issue of ceph (mon, osd) metadata that is generated by ceph and is needs to be persisted to local disk no matter what.  Not sure how much of that can be regenerated, i.e. is it possible to store all the config we need in configMaps such that when a pod is restarted it can regenerate all its metadata?  I doubt that's possible for things like OSD databases, but maybe others would be OK.

**Comment by bassam** [comment]
@travisn does this remove the need for dataDirHostPath completely?

**Comment by travisn** [comment]
yes, that is the hope, i just haven't been able to put together a prototype yet

**Comment by bassam** [comment]
would be great to get rid of dataDirHostPath in 0.6. Its confusing and with it not being the default it means we don't survive reboots by default :-(

**Comment by travisn** [comment]
yes, very much would like to make that happen, i'm working on it next. Also to consider is that local storage is still in alpha

**Comment by travisn** [comment]
I don't see it feasible to remove the `dataDirHostPath` at this time while local storage is in alpha. Nearly everything in the "what doesn't work" and "future features" sections are necessary for rook to effectively use the local storage on the mon and osd pods.

The documentation for [local storage](https://github.com/kubernetes-incubator/external-storage/tree/master/local-volume#feature-status) indicates the following issues while it is in alpha:
- "PVC binding does not consider pod scheduling requirements and may make suboptimal or incorrect decisions."
- "External provisioner cannot correctly detect capacity of mounts added after it has been started."

For OSDs to build on the raw devices, we would need these future features:
- "Local block devices as a volume source, with partitioning and fs formatting"
- "Pod accessing local raw block device"
- "Inline PV (use dedicated local disk as ephemeral storage)"
- "Dynamic provisioning for shared local persistent storage"

**Comment by DanKerns** [comment]
Is local storage in 1.9 viable? If not we should comment on the plan and see if we can get someting that would work for us.

**Comment by travisn** [comment]
@bassam I don't see how we can do this before 1.9, sound ok to move it out to 0.7?

**Comment by bassam** [comment]
@travisn given that local storage is in the external-storage repo (and not part of Kubernetes core), could we make these changes necessary to support rook scenarios?

**Comment by travisn** [comment]
@bassam it would appear that 1.9 which is just around the corner so I don't see how we could accelerate that very much.

**Comment by jbw976** [comment]
We discussed just simply setting a reasonable default for `dataDirHostPath` in the 0.6 milestone.  Maybe `/var/lib/rook`?  It is doubtful that just setting that default in the sample `rook-cluster.yaml` would make us any worse off than we are right now where Rook is not by default persistent.

**Comment by travisn** [comment]
@bassam per slack discussion, would suggest we move this out of 0.6

**Comment by kokhang** [comment]
@travisn @jbw976 @bassam @DanKerns: I need some guidance on how we want to proceed with this. 

I started to play with local-storage and its [external provisioner](https://github.com/kubernetes-incubator/external-storage/tree/master/local-volume).

There are a few things that makes the usage of local-storage not too trivial for Rook.
1- Local-storage is alpha on 1.8 and perhaps on 1.9. So users will have to enable it via the feature-gate. This involves updating kubelet and kube-api and restarting them.

2- The user will have to create and prepare directories in the nodes before hand to be used for local-storage. This involves creating the directory and mounting a temp fs ahead of time. See [here](https://github.com/kubernetes-incubator/external-storage/tree/master/local-volume#option-3-baremetal-environments). We can, perhaps, leverage the rook-agent to create this directory for us. So upon start, the rook-agent will create a directory to be used a local-storage on the node. Then, when a rook-cluster is created, Rook provisions the local-storage volume and uses it as the `dataDirHostPath`. But i would like to hear if there are other ways.

3- I got some bumps while trying the local-storage provisioner. But i think these can be fixed. Here are the issues i filed:
- https://github.com/kubernetes-incubator/external-storage/issues/462
- https://github.com/kubernetes-incubator/external-storage/issues/463

**Comment by kokhang** [comment]
OK.. after talking to Michelle, i got local-volume working.

I think the only limitation is the fact that local-volume is an alpha feature and has to be enable via feature-gate. Michelle mentioned that a realistic target for beta is 1.10.

One option is to support both local-volume and hostpath. Rook would use local-volume to persist the cluster metadata or fallback to using hostpath if local-volume is not enabled.

**Comment by kokhang** [comment]
Another issue found is that OSDs are deployed via a daemonset. Local-storage are just like any other persistent volume, where it doesnt work well with group of stateless pods.

So two options:
1- Create a statefulset and use anti-affinity to ensure it is deployed on all nodes
2-Create multiple replicaset (one of each node), each of size one. For every replicaset, we use anti-affinity to deploy it on all nodes

The disadvantage is that Rook has to watch for nodes in order to scale automatically as more nodes are added.

**Comment by jbw976** [comment]
@kokhang in the case where individual nodes are specified in the Cluster CRD storage spec, we already go the path of using a replicaset per node.  Taking option 2 from above could unify those 2 code paths, but I'd be even more interested in that option with the work I'm doing on adding/removing storage to the cluster.  In the design I'm writing, it's beneficial to add 1 node at a time to slowly handle rebalancing of data in the cluster.  If the operator has more control over when nodes are added in (which it can't do with a DaemonSet), then we can achieve that slow addition of nodes more easily.  What do you think?

**Comment by kokhang** [comment]
@jbw976 one thing that i forgot to say is that using a statefulset would allow Kubernetes to also manage the volume. Since the volume is backed by the node, this is probably not an issue. So yes, that makes sense.

**Comment by kokhang** [comment]
@bassam how do you advice to proceed on this? Currently `LocalVolume` is an alpha feature on 1.8 and possible on 1.9, which means it requires user intervention and kubelet restart to enable it.

In the mean time, we could alleviate some of the current issues. For example, we could have Rook delete the `dataDirHostPath` when a cluster is being deleted so that when the cluster is redeploy, it doesnt fail.

**Comment by sebastien-prudhomme** [comment]
@galexrt Any news on this feature now that local storage is in beta since Kubernetes 1.10

**Comment by galexrt** [comment]
This will probably be handled by #919.

**Comment by kfox1111** [comment]
unstale please

**Comment by mamoit** [comment]
Is the objective of this issue to use [raw block volume support](https://kubernetes.io/docs/concepts/storage/persistent-volumes/#raw-block-volume-support) which is beta in kubernetes 1.13?
This would allow for rolling upgrades in nodes where the data would be moved along with them, which is a must have in environments such as GKE and EKS, and maybe even (who knows), running ceph nodes in preemptible/spot instances. Live dangerously!

**Comment by travisn** [comment]
@mamoit Yes, in order to work independently from the host path, we will have a more complete design based on PVCs. It will be more general than just the local storage as described in this issue, but will use any storage class to create the PVCs, whether using the local storage provisioner, ebs, etc. This design is actively in progress and @jarrpa will have a design doc out there in the next few days. The work will be tracked with #919 rather than this issue since this is a subset of the more general design.

**Comment by mamoit** [comment]
@travisn Thank you very much for your prompt reply, I'll track #919.
If you need some stress testing let me know :)
